### PR TITLE
Add schema multi-version launch week blog post

### PR DESCRIPTION
**Pull request by andrew-farries**
## Summary

Direct link to post: https://xata-blog-pr-278.vercel.app/blog/multi-version-schema-migrations

### CODE REVIEW

File: schema-multiversion-launch-week.mdx [hunk 1]
```diff
@@ -0,0 +1,237 @@
+---
+title: Schema multi-version early access preview goes live today
```
Comment (alexfrancoeur) [inline schema-multiversion-launch-week.mdx:2]: I don't know if we should call it an early access preview as it rolls into the public beta for our serverless postgres offering. What do you think about just announcing this functionality and then talking about it as part of the general public beta or postgres?

Comment (andrew-farries) [reply schema-multiversion-launch-week.mdx:2]: Sure. For some reason I thought we were calling this an EAP. Happy to remove it.

Comment (andrew-farries) [reply schema-multiversion-launch-week.mdx]: Removed

File: schema-multiversion-launch-week.mdx [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+## Creating the schema
+
+We'll use the Xata UI to:
+
+- Create a simple `users` table with a `name` column
+- Populate the table with some random data
+
+In a new Xata database (with Postgres access enabled), we can create a new table in the Xata UI and populate it with some random data:
+
+![Create a new `users` table in the Xata UI and add a `name` column](images/schema-multiversion-launch-week/new-table.png)
```
Comment (alexfrancoeur) [inline schema-multiversion-launch-week.mdx:51]: Removing explicit UI references. Probably more of a nit than anything

```suggestion
We'll use Xata to:

- Create a simple `users` table with a `name` column
- Populate the table with some random data

In a new Xata database (with Postgres access enabled), we can create a new table in the app and populate it with some random data:

![Create a new `users` table in Xata and add a `name` column](images/schema-multiversion-launch-week/new-table.png)
```

File: schema-multiversion-launch-week.mdx [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+]
+```
+
+The SQL expression in the `up` field is used to migrate data from the old version of the `name` column that did not have the `NOT NULL` constraint to the new version of the column that does. This means the SQL expression needs to describe what to do with any `NULL` values that may already exist in the `name` field; here we are going to rewrite any `NULL`s to a placeholder value of `(unknown)`. Any non-`NULL` vales are copied over to the new version of the `name` column without modification.
+
+The `down` SQL expression does the opposite; it describes how values written to the new version of the `name` column will be migrated back to the old version of the column. Here we simply copy the value without modification.
+
+Arbitrary transformations can be applied to the data during the migration, allowing for complex data migrations to be performed as part of the schema migration.
+
+We can take this migration and run it via the Migration Editor in the Xata UI:
```
Comment (alexfrancoeur) [inline schema-multiversion-launch-week.mdx:83]:
```suggestion
We can take this migration and run it via the Migration Editor:
```

File: schema-multiversion-launch-week.mdx [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+> Once a migration is completed, it cannot be rolled back. Before completing a migration, make sure that all applications that depend on the old version of the schema have been updated to the new version so that the old version of the database schema is no longer required.
+
+Completing an in-progress migration can be done from various places in the Xata UI, including the migrations widget on the dashboard:
+
+![The migration can be completed from the Xata UI](images/schema-multiversion-launch-week/complete.png)
+
+Once completed, the table UI once again shows just one version of the `name` column in the `users` table:
+
+![After migration completion only the final version of the `name` column remains on the table](images/schema-multiversion-launch-week/table-editor-2.png)
```
Comment (alexfrancoeur) [inline schema-multiversion-launch-week.mdx:217]:
```suggestion
Completing an in-progress migration can be done from various places in Xata, including the migrations widget on the dashboard:

![The migration can be completed from Xata](images/schema-multiversion-launch-week/complete.png)

Once completed, the table shows just one version of the `name` column in the `users` table again:

![After migration completion only the final version of the `name` column remains on the table](images/schema-multiversion-launch-week/table-editor-2.png)
```

File: schema-multiversion-launch-week.mdx [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+## Rolling back a migration
+
+Because schema multi-version uses the expand/contract pattern to apply schema changes, rollbacks become much easier. The old version of the database schema was never removed so a rollback means simply removing the expanded parts of the database schema (the new version of the `users.name` column in our example).
+
+Like completing a migration, a migration can be rolled back from the migrations widget on the dashboard:
+
+![The migration can be rolled back from the Xata UI](images/schema-multiversion-launch-week/rollback.png)
```
Comment (alexfrancoeur) [inline schema-multiversion-launch-week.mdx:225]:
```suggestion
![The migration can be easily rolled back](images/schema-multiversion-launch-week/rollback.png)
```

File: schema-multiversion-launch-week.mdx [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+description: Schema multi-version, released today, addresses one of the most common pain points of application deployment - keeping your application code and database schema in sync. Starting today, database migrations in Xata can present two versions of your schema, old and new, to client applications allowing users to roll out new versions of their production applications with confidence.
+image:
+  src: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/schema-multi-version@2x.jpg
+  alt: Schema multi-version
+author: Andrew Farries
+authorEmail: andrew@xata.io
+date: 07-12-2024
+tags: ['open-source', 'postgres', 'schema', 'migrations', 'pgroll', 'fpSchemaMigrations']
+published: true
+slug: schema-multi-version-launch-week
```
Comment (alexfrancoeur) [inline schema-multiversion-launch-week.mdx:12]:
```suggestion
slug: multi-version-schema-migrations
```

**Review by alexfrancoeur — APPROVED** [review]
@andrew-farries @eemmiillyy really great blog! I made a few suggestions, most of which are minor, so I'll summarize the changes I'm suggesting for tomorrow.

* Rather than calling this feature schema multi-version, I'm suggestion we stay consistent across all entities (docs, app, web, brand) and call this capability multi-version schema migrations. What do you think?
* More of a nit than anything, but I removed references to our UI when I saw them. I like to think of Xata as one product, and referencing the UI explicitly always feels like an internal separate we make
* I updated the title + description and conclusion. For the title + description, I tried to simplify a bit. For the conclusion, I added in the demo video and an outro similar to todays blog post. I could use @CezzaineZ 's input here though.

If these changes are mostly good with you, ✅ from me!

File: schema-multiversion-launch-week.mdx [hunk 7]
```diff
@@ -0,0 +1,237 @@
+---
+title: Schema multi-version goes live today
+description: Schema multi-version, released today, addresses one of the most common pain points of application deployment - keeping your application code and database schema in sync. Starting today, database migrations in Xata can present two versions of your schema, old and new, to client applications allowing users to roll out new versions of their production applications with confidence.
+image:
+  src: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/schema-multi-version@2x.jpg
```
[the remark this answers is not in the document]
Comment (andrew-farries) [reply schema-multiversion-launch-week.mdx:5]: done

File: schema-multiversion-launch-week.mdx [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+published: true
+slug: schema-multi-version-launch-week
+ogImage: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/schema-multi-version@2x.jpg
+---
+
+Today we are excited to announce schema multi-version in Xata.
+
+Schema multi-version tackles one of the most painful aspects of application deployment: keeping your application code and database schema in sync.
+
+Schema multi-version means that both the old and new versions of your database schema are live and accessible during a schema migration. This greatly simplifies deployment of your applications as both the currently deployed and the new version of your application see the version of the database schema that they expect. And because both versions of the database schema are live at the same time, rollbacks become trivial.
```
[the remark this answers is not in the document]
Comment (andrew-farries) [reply schema-multiversion-launch-week.mdx:30]: Went with more-or-less what you suggested.

**Comment by andrew-farries** [comment]
Thanks for the detailed review @alexfrancoeur 🙏 ❤️ 

I've applied all the suggestions.

**Comment by andrew-farries** [comment]
> Rather than calling this feature schema multi-version, I'm suggestion we stay consistent across all entities (docs, app, web, brand) and call this capability multi-version schema migrations. What do you think?

"multi-version schema migrations" is a much better term; I've used it everywhere in the blog now.

> More of a nit than anything, but I removed references to our UI when I saw them. I like to think of Xata as one product, and referencing the UI explicitly always feels like an internal separate we make

👍 Agree on this. References to UI have been removed.

File: schema-multiversion-launch-week.mdx [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+ogImage: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/schema-multi-version@2x.jpg
+---
+
+Today we are excited to announce multi-version schema migrations in Xata 🎉
+
+Multi-version schema migrations tackle one of the most painful aspects of application deployment: keeping your application code and database schema in sync.
+
+Multi-version schema migrations allow both the old and new versions of your database schema to be live and accessible during a schema migration. This greatly simplifies deployment of your applications as both the currently deployed and the new version of your application see the version of the database schema that they expect. And because both versions of the database schema are live at the same time, rollbacks become trivial.
+
+Behind the scenes, Xata does the heavy-lifting to make this work by presenting different versioned views of your application schema to each version of your application and transparently migrating data between the old and new versions of your database schema as data is written to each version.
```
Comment (CezzaineZ) [inline schema-multiversion-launch-week.mdx:32]:
```suggestion
Behind the scenes, Xata does the heavy-lifting to make this work by presenting different versioned views of your application schema to each version of your application. It then transparently migrates data between the old and new versions of your database schema as data is being written to each version.
```

File: schema-multiversion-launch-week.mdx [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+First, we'll create a table in a fresh database, populate it with some data and then create a migration to add a `NOT NULL` constraint to a column in the table. We'll see how multi-version schema migrations allow two versions of the database schema, one with the constraint and one without, to be made available to applications.
+
+## Creating the schema
+
+We'll use Xata to:
+
+- Create a simple `users` table with a `name` column
+- Populate the table with some random data
+
+In a new Xata database (with Postgres access enabled), we can create a new table in the app and populate it with some random data:
```
Comment (CezzaineZ) [inline schema-multiversion-launch-week.mdx:59]:
```suggestion
In a new Xata database (with Postgres access enabled), we can create a new table in the applicatioon and populate it with some random data:
```

File: schema-multiversion-launch-week.mdx [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+      "column": "name",
+      "nullable": false,
+      "up": "(SELECT CASE WHEN name IS NULL THEN '(unknown)' ELSE name END)",
+      "down": "name"
+    }
+  }
+]
+```
+
+The SQL expression in the `up` field is used to migrate data from the old version of the `name` column that did not have the `NOT NULL` constraint to the new version of the column that does. This means the SQL expression needs to describe what to do with any `NULL` values that may already exist in the `name` field; here we are going to rewrite any `NULL`s to a placeholder value of `(unknown)`. Any non-`NULL` vales are copied over to the new version of the `name` column without modification.
```
Comment (CezzaineZ) [inline schema-multiversion-launch-week.mdx:87]:
```suggestion
The SQL expression in the `up` field is used to migrate data from the old version of the `name` column that did not have the `NOT NULL` constraint to the new version of the column that does. This means the SQL expression needs to describe what to do with any `NULL` values that may already exist in the `name` field; here we are going to rewrite any `NULL`s to a placeholder value of `(unknown)`. Any non-`NULL` values are copied over to the new version of the `name` column without modification.
```

File: schema-multiversion-launch-week.mdx [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+
+We can take this migration and run it via the Migration Editor:
+
+![Start a migration to add a `NOT NULL` constraint to the `users` table](images/schema-multiversion-launch-week/start-migration.png)
+
+Once the migration is started, the table UI shows that we have two versions of the `name` column in the `users` table, one with the `NOT NULL` constraint and one without (this is the [expand/contract pattern](https://blog.thepete.net/blog/2023/12/05/expand/contract-making-a-breaking-change-without-a-big-bang/) in practice):
+
+![The table editor shows the two versions of the `name` field, one with the `NOT NULL` constraint and one without](images/schema-multiversion-launch-week/table-editor.png)
+
+The new version of the `name` column has been filled with the data from the old version of the column. Had there been any `NULL` values in the column, these would have been rewritten by the `up` data migration we specified in the migration file. The two versions of the `name` column are always kept in sync for the duration of the migration; as data is written into either one it is migrated to the other using the `up` and `down` SQL expressions from the migration file.
```
Comment (CezzaineZ) [inline schema-multiversion-launch-week.mdx:101]:
```suggestion
The new version of the `name` column has been filled with the data from the old version of the column. Had there been any `NULL` values in the column, these would have been rewritten by the `up` data migration we specified in the migration file. The two versions of the `name` column are always kept in sync for the duration of the migration; as data is written into either one, it is migrated to the other using the `up` and `down` SQL expressions from the migration file.
```

File: schema-multiversion-launch-week.mdx [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+The migration is now in progress and the two versions of the database schema are live and available to applications.
+
+## Connecting to each schema version
+
+In order to be able to take advantage of multi-version schema migrations, applications need to be able to connect to the correct version of the database schema. This is done using the Postgres [search_path](https://www.postgresql.org/docs/current/ddl-schemas.html#DDL-SCHEMAS-PATH) setting. The `search_path` setting determines which schema will be searched for non-schema-qualified table accesses. By setting the `search_path` appropriately, client applications can access the version of the database schema that is compatible with the deployed application code.
```
Comment (CezzaineZ) [inline schema-multiversion-launch-week.mdx:107]:
```suggestion
In order to be able to take advantage of multi-version schema migrations, applications need to be able to connect to the correct version of the database schema. This is done using the Postgres [search_path](https://www.postgresql.org/docs/current/ddl-schemas.html#DDL-SCHEMAS-PATH) setting. The `search_path` setting determines which schema will be searched for non-schema-qualified table access. By setting the `search_path` appropriately, client applications can access the version of the database schema that is compatible with the deployed application code.
```

File: schema-multiversion-launch-week.mdx [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+-- |--------------------------|----------------------|
+-- | rec_cq7v4a18fe4f6dj1qj8g | yuck frantically     |
+-- | rec_cq7v4a18fe4f6dj1qjf0 | worriedly disclose   |
+-- | rec_cq7v4a18fe4f6dj1qjj0 | weaponize well-to-do |
+-- | rec_cq7v88tqrj67kk4bk9l0 | (unknown)            |
+-- | rec_cq7v4a18fe4f6dj1qjcg | tensely drafty       |
+-- ...
+```
+
+In the old version of the schema the `name` field allows `NULL` values, so the `INSERT` statement succeeeds and reads from the table show the new row with a `NULL` value in the `name` field. When the same data is read in the new version of the schema the same row has the placeholder value of `(unknown)` that was defined in the migration.
```
Comment (CezzaineZ) [inline schema-multiversion-launch-week.mdx:211]:
```suggestion
In the old version of the schema, the `name` field allows `NULL` values, so the `INSERT` statement succeeds and reads from the table showing the new row with a `NULL` value in the `name` field. When the same data is read in the new version of the schema, the same row has the placeholder value of `(unknown)` that was defined in the migration.
```

**Review by CezzaineZ — COMMENTED** [review]
Left a few suggested amends but otherwise LGTM :)

**Comment by andrew-farries** [comment]
Thanks for the review @CezzaineZ 🙏 I've applied the suggestions.

File: schema-multiversion-launch-week.mdx [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+
+# Conclusion
+
+Multi-version schema migrations simplify one of the most common pain points when deploying applications: keeping the deployed versions of your applications and the database schema in sync.
+
+This new feature allows you to deploy your applications and make database schema changes independently and with confidence.
+
+We are excited to share this with you and look forward to your feedback! Want to see multi-version schema migrations in action? Check out this quick demo video.
+
+<ArticleVideo platform="youtube" src="https://www.youtube.com/embed/uG4q_B50e1Q?si=2BCEzMeTct3HVsHn" />
```
Comment (alexfrancoeur) [inline schema-multiversion-launch-week.mdx:249]:
```suggestion
<ArticleVideo platform="youtube" src="https://www.youtube.com/embed/TYoW32lF6Xs?si=QiRWjNknkSxt2Mm8" />
```

**Pull request merged**